File: dsa-problems/Tries/solutions/javascript/TRI-014-longest-word-by-k-prefixes.js
const readline = require("readline");

class TrieNode {
  constructor() {
    this.children = new Map();
    this.isEnd = false;
  }
}

class Solution {
  constructor() {
    this.root = new TrieNode();
  }

  insert(word) {
    let node = this.root;
    for (const char of word) {
      if (!node.children.has(char)) {
        node.children.set(char, new TrieNode());
      }
      node = node.children.get(char);
    }
    node.isEnd = true;
  }

  countPrefixes(word) {
    let node = this.root;
    let count = 0;

    for (let i = 0; i < word.length; i++) {
      node = node.children.get(word[i]);
      if (i < word.length - 1 && node.isEnd) {
        count++;
      }
    }

    return count;
  }

  longestWordWithKPrefixes(words, k) {
    // Build trie
    for (const word of words) {
      this.insert(word);
    }

    let result = "";
    let maxLen = 0;

    // Check each word
    for (const word of words) {
      const prefixCount = this.countPrefixes(word);
      if (prefixCount >= k) {
        if (word.length > maxLen || (word.length === maxLen && word < result)) {
          result = word;
          maxLen = word.length;
        }
      }
    }

    return result;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const lines = [];
rl.on("line", (line) => {
  lines.push(line);
}).on("close", () => {
  const [n, k] = lines[0].split(" ").map(Number);
  const words = [];
  for (let i = 1; i <= n; i++) {
    words.push(lines[i].trim());
  }

  const solution = new Solution();
  const result = solution.longestWordWithKPrefixes(words, k);

  console.log(result);
});

File: dsa-problems/Tries/solutions/cpp/TRI-014-longest-word-by-k-prefixes.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    bool isEnd = false;
};

class Solution {
private:
    TrieNode* root;

    void insert(const string& word) {
        TrieNode* node = root;
        for (char c : word) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
        }
        node->isEnd = true;
    }

    int countPrefixes(const string& word) {
        TrieNode* node = root;
        int count = 0;

        for (int i = 0; i < word.length(); i++) {
            node = node->children[word[i]];
            if (i < word.length() - 1 && node->isEnd) {
                count++;
            }
        }

        return count;
    }

public:
    Solution() { root = new TrieNode(); }

    string longestWordWithKPrefixes(vector<string>& words, int k) {
        // Build trie
        for (const string& word : words) {
            insert(word);
        }

        string result = "";
        int maxLen = 0;

        // Check each word
        for (const string& word : words) {
            int prefixCount = countPrefixes(word);
            if (prefixCount >= k) {
                if (word.length() > maxLen ||
                    (word.length() == maxLen && word < result)) {
                    result = word;
                    maxLen = word.length();
                }
            }
        }

        return result;
    }
};

int main() {
    int n, k;
    cin >> n >> k;
    cin.ignore();

    vector<string> words(n);
    for (int i = 0; i < n; i++) {
        getline(cin, words[i]);
    }

    Solution solution;
    string result = solution.longestWordWithKPrefixes(words, k);

    cout << result << endl;

    return 0;
}
